core/src/main/java/io/confluent/rest/filters/CsrfTokenProtectionFilter.java

```diff
@@ -0,0 +1,182 @@
+  public static final String INVALID_REQUESTER_MESSAGE = "Missing user session identifier";
+  private static final Set<String> METHODS_TO_IGNORE;
+
+  private String csrfTokenEndpoint = RestConfig.CSRF_PREVENTION_TOKEN_FETCH_ENDPOINT_DEFAULT;
+  private int csrfTokenExpiration = RestConfig.CSRF_PREVENTION_TOKEN_EXPIRATION_MINUTES_DEFAULT;
+  private int csrfTokenMaxEntries = RestConfig.CSRF_PREVENTION_TOKEN_MAX_ENTRIES_DEFAULT;
+  private LoadingCache<String, String> tokenSupplier;
+
+  static {
+    HashSet<String> mti = new HashSet();
+    mti.add("GET");
+    mti.add("OPTIONS");
+    mti.add("HEAD");
+    METHODS_TO_IGNORE = Collections.unmodifiableSet(mti);
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+    if (filterConfig.getInitParameter(RestConfig.CSRF_PREVENTION_TOKEN_FETCH_ENDPOINT) != null) {
+      this.csrfTokenEndpoint =
+          filterConfig.getInitParameter(RestConfig.CSRF_PREVENTION_TOKEN_FETCH_ENDPOINT);
+    }
+
+    if (filterConfig.getInitParameter(RestConfig.CSRF_PREVENTION_TOKEN_EXPIRATION_MINUTES)
+        != null) {
+      this.csrfTokenExpiration =
+          Integer.parseInt(
+              filterConfig.getInitParameter(RestConfig.CSRF_PREVENTION_TOKEN_EXPIRATION_MINUTES));
+    }
+
+    if (filterConfig.getInitParameter(RestConfig.CSRF_PREVENTION_TOKEN_MAX_ENTRIES)
+        != null) {
+      this.csrfTokenMaxEntries =
+          Integer.parseInt(
+              filterConfig.getInitParameter(RestConfig.CSRF_PREVENTION_TOKEN_MAX_ENTRIES));
+    }
+
+    this.tokenSupplier =
+        CacheBuilder.newBuilder()
+            .expireAfterWrite(csrfTokenExpiration, TimeUnit.MINUTES)
+            .maximumSize(csrfTokenMaxEntries)
+            .build(
+                new CacheLoader<String, String>() {
+                  @Override
+                  public String load(String key) throws Exception {
+                    return UUID.randomUUID().toString();
+                  }
+                });
+  }
+
+  @Override
+  public void doFilter(
+      ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
+      throws IOException, ServletException {
+
+    HttpServletRequest httpServletRequest = (HttpServletRequest) servletRequest;
+    HttpServletResponse httpServletResponse = (HttpServletResponse) servletResponse;
+    String requestedBy = httpServletRequest.getHeader(REQUESTED_BY);
+
+    // Always check for request's METHOD before checking for CSRF_ENDPOINT
+    if (!METHODS_TO_IGNORE.contains(httpServletRequest.getMethod())) {
+      // For guarded methods 'requestedBy' is required
+      if (Strings.isNullOrEmpty(requestedBy)) {
+        log.error("(Cross site request forgery): Missing user session identifier");
+        throw new BadRequestException(INVALID_REQUESTER_MESSAGE);
+      }
+
+      // Check if the token in request header indeed belongs to requested user
+      if (!tokenSupplier
+          .getUnchecked(requestedBy)
```

Comment:
should we change here to getIfPresent() s.t. if it's an invalid token, we don't load it in the cache?

Reply:
Great catch